File: kernel/livepatch/guard.rs
// SPDX-License-Identifier: GPL-2.0-only
// Copyright (C) 2026 rezky_nightky

//! Kernel livepatch safety guards — prevent kernel panics.
//!
//! Every operation that touches kernel execution flow must pass
//! through a guard. Guards check preconditions, wrap dangerous
//! operations in error-handling, and return structured errors.
//!
//! ## Safety Principles
//!
//!   1. Check BEFORE acting — never act and then discover failure
//!   2. Return errors, never panic — userspace handles rejection
//!   3. Isolate dangerous code — stop_machine lives in a guarded scope
//!   4. Recover gracefully — any unexpected state → structured error
//!
//! ## Guard Lifecycle
//!
//!   Module Init:
//!     preflight_checks() → Ok(()) or refuse to load
//!
//!   Patch Request:
//!     validate_target() → verify symbol + address
//!     guarded_stop_machine() → atomic apply with error capture
//!
//!   Module Unload:
//!     require_no_active_patches() → refuse unload if patches active

#![allow(dead_code)] // Architectural skeleton — compiled inside real kernel tree

use alloc::string::String;
use alloc::vec::Vec;

// ============================================================================
//  Pre-flight Guard — runs at module init
// ============================================================================

/// Result of a pre-flight check.
#[derive(Clone, Debug)]
pub struct PreflightResult {
    /// All checks passed.
    pub ok: bool,
    /// List of failures (empty if ok).
    pub failures: Vec<PreflightFailure>,
}

#[derive(Clone, Debug)]
pub struct PreflightFailure {
    /// Which check failed.
    pub check: &'static str,
    /// Why it failed.
    pub reason: String,
    /// Is this fatal? (true = module must abort load)
    pub fatal: bool,
}

/// Run all pre-flight checks at module init.
///
/// If ANY fatal check fails, the module MUST refuse to load.
/// Userspace receives the structured failure reason via dmesg.
pub fn preflight_checks() -> PreflightResult {
    let mut failures = Vec::new();

    // ── Check 1: CONFIG_LIVEPATCH ──
    // In real kernel code:
    //   if !cfg!(CONFIG_LIVEPATCH) {
    //       failures.push(PreflightFailure {
    //           check: "CONFIG_LIVEPATCH",
    //           reason: "Livepatch support not compiled into this kernel".into(),
    //           fatal: true,
    //       });
    //   }
    check_config("LIVEPATCH", true, &mut failures);

    // ── Check 2: CONFIG_FUNCTION_TRACER ──
    check_config("FUNCTION_TRACER", true, &mut failures);

    // ── Check 3: CONFIG_MODULES ──
    check_config("MODULES", true, &mut failures);

    // ── Check 4: Running in safe environment ──
    // Some container runtimes restrict kernel code modification.
    // Check /proc/1/environ for container indicators.
    check_container_safety(&mut failures);

    // ── Check 5: Kernel lockdown ──
    // If kernel lockdown is in 'confidentiality' mode, livepatch is blocked.
    // Read /sys/kernel/security/lockdown — if "confidentiality", refuse.
    check_lockdown_mode(&mut failures);

    PreflightResult {
        ok: failures.iter().all(|f| !f.fatal),
        failures,
    }
}

fn check_config(name: &str, required: bool, failures: &mut Vec<PreflightFailure>) {
    // In real kernel code: IS_ENABLED(CONFIG_{name})
    let enabled = true; // stub — real code checks kernel config
    if required && !enabled {
        failures.push(PreflightFailure {
            check: "CONFIG_LIVEPATCH",
            reason: format!("CONFIG_{name} is not enabled in this kernel"),
            fatal: true,
        });
    }
}

fn check_container_safety(failures: &mut Vec<PreflightFailure>) {
    // In real kernel code:
    //   - Check if init process is systemd-nspawn or docker-init
    //   - Check cgroup namespace for container indicators
    //   - Check seccomp profile for restricted syscalls
    //
    // If running in a restricted container that blocks kernel code
    // modification, refuse to load.
    //
    // This is a WARNING, not fatal — containers with CAP_SYS_MODULE
    // can still use livepatch.

    // Stub: assume safe
    let _ = failures;
}

fn check_lockdown_mode(failures: &mut Vec<PreflightFailure>) {
    // In real kernel code, read the lockdown LSM state:
    //
    //   let lockdown = security_locked_down(LOCKDOWN_KERNEL_MODULE);
    //   if lockdown == LOCKDOWN_CONFIDENTIALITY_MAX {
    //       failures.push(PreflightFailure {
    //           check: "kernel_lockdown",
    //           reason: "Kernel lockdown is in confidentiality mode".into(),
    //           fatal: true,
    //       });
    //   } else if lockdown == LOCKDOWN_INTEGRITY_MAX {
    //       failures.push(PreflightFailure {
    //           check: "kernel_lockdown",
    //           reason: "Kernel lockdown is in integrity mode — some patches may be restricted".into(),
    //           fatal: false, // non-fatal warning
    //       });
    //   }

    let _ = failures;
}

// ============================================================================
//  Target Validation Guard — runs before every patch
// ============================================================================

/// Validate a patch target before attempting to apply.
pub fn validate_target(symbol_name: &str, _target_address: usize) -> Result<(), String> {
    // In real kernel code:
    //
    //   let addr = kallsyms_lookup_name(symbol_name);
    //   if addr.is_null() {
    //       return Err(format!("Symbol '{}' not found in kallsyms", symbol_name));
    //   }
    //
    //   // Check symbol is in .text section (only code can be livepatched)
    //   if !is_text_symbol(addr) {
    //       return Err(format!("Symbol '{}' is not in .text section", symbol_name));
    //   }
    //
    //   // Check symbol is not already patched
    //   if is_currently_patched(addr) {
    //       return Err(format!("Symbol '{}' is already patched", symbol_name));
    //   }
    //
    //   // Check symbol is not blacklisted (NOKPROBE, NOTRACE)
    //   if is_blacklisted(addr) {
    //       return Err(format!("Symbol '{}' is marked NOTRACE/NOKPROBE", symbol_name));
    //   }

    let _ = symbol_name;
    Ok(())
}

// ============================================================================
//  Atomic Execution Guard — wraps stop_machine safely
// ============================================================================

/// Result of a guarded atomic operation.
pub struct GuardedExecResult {
    pub success: bool,
    pub error: Option<String>,
    /// If the pre-patch state was saved, it's here for rollback.
    pub saved_state: Option<SavedPatchState>,
}

/// Saved state before a patch — enables safe rollback.
pub struct SavedPatchState {
    pub symbol_name: String,
    pub old_address: usize,
    pub old_handler: usize,
}

/// Execute a guarded stop_machine operation.
///
/// This is the CRITICAL SECTION of livepatch. Every failure path
/// must be handled — the kernel must NEVER oops here.
///
/// Safety contract:
///   1. text_mutex MUST be acquired before calling this
///   2. stop_machine MUST complete (no partial application)
///   3. On ANY error, restore original state
///   4. Always release text_mutex (even on panic path)
pub fn guarded_stop_machine(
    symbol_name: &str,
    old_address: usize,
    new_address: usize,
) -> GuardedExecResult {
    // In real kernel code:
    //
    //   // Step 0: Save state for rollback
    //   let saved = SavedPatchState {
    //       symbol_name: symbol_name.into(),
    //       old_address,
    //       old_handler: read_ftrace_handler(old_address),
    //   };
    //
    //   // Step 1: Acquire text_mutex
    //   let mutex_result = mutex_lock_interruptible(&text_mutex);
    //   if mutex_result != 0 {
    //       return GuardedExecResult {
    //           success: false,
    //           error: Some("Failed to acquire text_mutex".into()),
    //           saved_state: None,
    //       };
    //   }
    //
    //   // Step 2: stop_machine with error capture
    //   //         The closure runs on a single CPU while all others are halted.
    //   let result = stop_machine(|_data| {
    //       // SAFETY: all other CPUs are halted — safe to modify code.
    //       //
    //       // Try to install ftrace handler. If this fails, we MUST
    //       // report the error — the stop_machine callback returns a
    //       // negative errno on failure.
    //       match install_ftrace_handler(old_address, new_address) {
    //           Ok(()) => 0,
    //           Err(e) => {
    //               pr_err!("zenvecha: ftrace install failed for {}: {}", symbol_name, e);
    //               -EIO
    //           }
    //       }
    //   }, NULL, NULL);
    //
    //   // Step 3: Always release text_mutex (even on failure)
    //   mutex_unlock(&text_mutex);
    //
    //   // Step 4: Interpret result
    //   if result == 0 {
    //       // Verify immediately
    //       let actual = read_ftrace_handler(old_address);
    //       if actual == new_address {
    //           GuardedExecResult {
    //               success: true,
    //               error: None,
    //               saved_state: Some(saved),
    //           }
    //       } else {
    //           // Mismatch — something went wrong silently
    //           GuardedExecResult {
    //               success: false,
    //               error: Some(format!("Handler mismatch: expected {}, got {}", new_address, actual)),
    //               saved_state: Some(saved),
    //           }
    //       }
    //   } else {
    //       GuardedExecResult {
    //           success: false,
    //           error: Some(format!("stop_machine failed with error {}", result)),
    //           saved_state: None,
    //       }
    //   }

    GuardedExecResult {
        success: true,
        error: None,
        saved_state: Some(SavedPatchState {
            symbol_name: symbol_name.into(),
            old_address,
            old_handler: old_address,
        }),
    }
}

// ============================================================================
//  Unload Guard — prevents unsafe module removal
// ============================================================================

/// Check if it's safe to unload the module.
///
/// Returns Ok(()) if no active patches exist.
/// Returns Err with a list of active patches that must be reverted first.
pub fn require_no_active_patches() -> Result<(), Vec<String>> {
    // In real kernel code:
    //   let active = list_active_patches();
    //   if !active.is_empty() {
    //       return Err(active);
    //   }
    Ok(())
}

File: kernel/guard.c
// SPDX-License-Identifier: GPL-2.0-only
// Copyright (C) 2026 rezky_nightky

//! Kernel livepatch safety guards — prevent kernel panics.
//!
//! Every operation that touches kernel execution flow must pass
//! through a guard. Guards check preconditions, wrap dangerous
//! operations in error-handling, and return structured errors.
//!
//! ## Safety Principles
//!
//!   1. Check BEFORE acting — never act and then discover failure
//!   2. Return errors, never panic — userspace handles rejection
//!   3. Isolate dangerous code — stop_machine lives in a guarded scope
//!   4. Recover gracefully — any unexpected state → structured error
//!
//! ## Six-Gate Safety Protocol
//!
//! Every patch must pass six gates before it is applied:
//!
//!   Gate 1: CONFIG_LIVEPATCH compiled in (compile-time check)
//!   Gate 2: CONFIG_FUNCTION_TRACER compiled in (compile-time)
//!   Gate 3: CONFIG_MODULES compiled in (compile-time)
//!   Gate 4: RuntimeRisk = low (read /proc/zenvecha/semantic.runtime_risk)
//!   Gate 5: No active patch on this symbol (check internal state)
//!   Gate 6: Symbol exists in our target registry (kallsyms-equivalent)
//!
//! Gates 1-3 are evaluated at module init (preflight). If any fails,
//! the module refuses to load entirely.
//! Gates 4-6 are evaluated at every apply request. If any fails,
//! the apply is rejected with a structured error message written
//! to /proc/zenvecha/livepatch/status.

#define pr_fmt(fmt) "zenvecha: " fmt

#include <linux/kernel.h>
#include <linux/string.h>
#include <linux/module.h>
#include <linux/atomic.h>
#include <linux/stop_machine.h>

#include "zenvecha.h"

/* Internal state — single-slot patch (one active patch at a time).
 *
 * For the PoC skeleton, only one patch is allowed at any time.
 * Production multi-patch support would require a hash table of
 * (symbol → patch_state) entries. */
static atomic_t zenvecha_patch_active = ATOMIC_INIT(0);
static char zenvecha_active_symbol[64];

bool zenvecha_patch_is_active(void)
{
	return atomic_read(&zenvecha_patch_active) != 0;
}

static bool symbol_is_known(const char *name)
{
	if (!name)
		return false;

	/* For the PoC skeleton, only zenvecha_dummy_func is a known
	 * patchable target. Production code would call kallsyms_lookup_name
	 * here — but that symbol is unexported since 5.7. The proper
	 * production approach is the kprobe-trick (register a kprobe with
	 * .symbol_name, read .addr, unregister). For now, match our own
	 * module symbol. */
	return strcmp(name, zenvecha_dummy_target_name()) == 0;
}

// ── Pre-flight Guard — runs at module init ─────────────────────────────

struct preflight_result zenvecha_preflight(void)
{
	struct preflight_result r = { .ok = true };

	/* Gate 1: CONFIG_LIVEPATCH */
#if !IS_ENABLED(CONFIG_LIVEPATCH)
	r.ok = false;
	r.fatal_check = "CONFIG_LIVEPATCH";
	r.fatal_reason = "Livepatch support not compiled into this kernel";
	return r;
#endif

	/* Gate 2: CONFIG_FUNCTION_TRACER */
#if !IS_ENABLED(CONFIG_FUNCTION_TRACER)
	r.ok = false;
	r.fatal_check = "CONFIG_FUNCTION_TRACER";
	r.fatal_reason = "Function tracer support not compiled into this kernel";
	return r;
#endif

	/* Gate 3: CONFIG_MODULES */
#if !IS_ENABLED(CONFIG_MODULES)
	r.ok = false;
	r.fatal_check = "CONFIG_MODULES";
	r.fatal_reason = "Module loader not compiled into this kernel";
	return r;
#endif

	/* All compile-time gates passed. */
	pr_info("preflight: CONFIG_LIVEPATCH + FUNCTION_TRACER + MODULES ok\n");
	return r;
}

// ── Target Validation Guard — runs before every patch ──────────────────

int zenvecha_validate_target(const char *symbol_name)
{
	/* Gate 4: RuntimeRisk must be "low" */
	if (!zenvecha_runtime_risk_is_low()) {
		pr_warn("gate 4 FAIL: runtime_risk=%s (must be 'low')\n",
			zenvecha_runtime_risk);
		return -EPERM;
	}

	/* Gate 5: No active patch */
	if (zenvecha_patch_is_active()) {
		pr_warn("gate 5 FAIL: active patch on '%s'\n",
			zenvecha_active_symbol);
		return -EBUSY;
	}

	/* Gate 6: Symbol exists in our target registry */
	if (!symbol_is_known(symbol_name)) {
		pr_warn("gate 6 FAIL: unknown symbol '%s'\n", symbol_name);
		return -ENOENT;
	}

	return 0;
}

// ── Atomic Execution Guard — stop_machine callback ─────────────────────

struct patch_payload {
	const char *symbol_name;
	int error; /* 0 = success, negative errno on failure */
};

static int apply_patch_atomic(void *data)
{
	struct patch_payload *p = data;

	/* All other CPUs are halted by stop_machine. We are now in the
	 * single critical section where it is safe to modify execution
	 * state.
	 *
	 * Skeleton: flip the dummy function's atomic flag.
	 * Production: install ftrace handler at the target address. */
	zenvecha_dummy_set_patched(true);
	p->error = 0;
	return 0;
}

static int revert_patch_atomic(void *data)
{
	struct patch_payload *p = data;

	zenvecha_dummy_set_patched(false);
	p->error = 0;
	return 0;
}

// ── Guarded apply / revert entry points ────────────────────────────────

int zenvecha_guarded_apply(const char *symbol_name)
{
	struct patch_payload payload = { .symbol_name = symbol_name };
	int ret;

	/* Run gates 4-6 */
	ret = zenvecha_validate_target(symbol_name);
	if (ret)
		return ret;

	/* Atomic apply via stop_machine — halts ALL other CPUs */
	ret = stop_machine(apply_patch_atomic, &payload, NULL);
	if (ret) {
		pr_err("stop_machine(apply) failed: %d\n", ret);
		return ret;
	}
	if (payload.error) {
		pr_err("apply callback failed: %d\n", payload.error);
		return payload.error;
	}

	/* Commit state */
	atomic_set(&zenvecha_patch_active, 1);
	strncpy(zenvecha_active_symbol, symbol_name,
		sizeof(zenvecha_active_symbol) - 1);
	zenvecha_active_symbol[sizeof(zenvecha_active_symbol) - 1] = '\0';

	pr_info("patch applied: %s (42 → 99)\n", symbol_name);
	return 0;
}

int zenvecha_guarded_revert(void)
{
	struct patch_payload payload = { .symbol_name = NULL };
	int ret;

	if (!zenvecha_patch_is_active()) {
		pr_info("revert: no active patch\n");
		return 0;
	}

	ret = stop_machine(revert_patch_atomic, &payload, NULL);
	if (ret) {
		pr_err("stop_machine(revert) failed: %d\n", ret);
		return ret;
	}

	pr_info("patch reverted: %s (99 → 42)\n", zenvecha_active_symbol);
	atomic_set(&zenvecha_patch_active, 0);
	zenvecha_active_symbol[0] = '\0';

	return 0;
}
